### PR TITLE
Allow Spack to uninstall external extensions without permissions

### LINKED ISSUE #11980: Unable to uninstall external package
On the latest develop, I tried uninstalling all of my packages, and Spack hung on an external `jdk` installation. It appears that Spack is unable to uninstall extendable packages (or any package that extends them) when that external package is installed in a directory the user does not have permission to access.

### Steps to reproduce the issue

Add the following to `packages.yaml`:
```yaml
packages:
    jdk:
        paths:
            jdk@10.0.1_10: /Library/Java/JavaVirtualMachines/jdk-10.0.1.jdk/Contents/Home
        buildable: False
```
Then try to install and uninstall the package:
```console
$ spack install commons-lang  # extends jdk
$ spack uninstall commons-lang
```

### Error Message

```console
$ spack -d uninstall commons-lang
...
Traceback (most recent call last):
  File "/Users/Adam/spack/bin/spack", line 48, in <module>
    sys.exit(spack.main.main())
  File "/Users/Adam/spack/lib/spack/spack/main.py", line 697, in main
    return _invoke_command(command, parser, args, unknown)
  File "/Users/Adam/spack/lib/spack/spack/main.py", line 447, in _invoke_command
    return_val = command(parser, args)
  File "/Users/Adam/spack/lib/spack/spack/cmd/uninstall.py", line 336, in uninstall
    args, spack.cmd.parse_specs(args.packages) if args.packages else [any])
  File "/Users/Adam/spack/lib/spack/spack/cmd/uninstall.py", line 326, in uninstall_specs
    do_uninstall(env, uninstall_list, args.force)
  File "/Users/Adam/spack/lib/spack/spack/cmd/uninstall.py", line 224, in do_uninstall
    item.do_uninstall(force=force)
  File "/Users/Adam/spack/lib/spack/spack/package.py", line 2017, in do_uninstall
    Package.uninstall_by_spec(self.spec, force)
  File "/Users/Adam/spack/lib/spack/spack/package.py", line 1997, in uninstall_by_spec
    spack.hooks.pre_uninstall(spec)
  File "/Users/Adam/spack/lib/spack/spack/hooks/__init__.py", line 54, in __call__
    hook(*args, **kwargs)
  File "/Users/Adam/spack/lib/spack/spack/hooks/extensions.py", line 16, in pre_uninstall
    view = YamlFilesystemView(target, spack.store.layout)
  File "/Users/Adam/spack/lib/spack/spack/filesystem_view.py", line 204, in __init__
    mkdirp(os.path.dirname(projections_path))
  File "/Users/Adam/spack/lib/spack/llnl/util/filesystem.py", line 436, in mkdirp
    raise e
  File "/Users/Adam/spack/lib/spack/llnl/util/filesystem.py", line 431, in mkdirp
    os.makedirs(path)
  File "/Users/Adam/miniconda3/lib/python3.7/os.py", line 221, in makedirs
    mkdir(name, mode)
PermissionError: [Errno 13] Permission denied: '/Library/Java/JavaVirtualMachines/jdk-10.0.1.jdk/Contents/Home/.spack'
```

### Information on your system

This occurred on macOS Mojave 10.14.5.

It appears that this bug only occurs with extensions, and only when an extension is actually installed. This bug is not exclusive to `jdk`, it can be reproduced with other packages as well. For example, add `libsigsegv` to `packages.yaml` in a directory lacking permissions. Then edit `libsigsegv` to add `extendable = True`, and edit `m4` to add `extends('libsigsegv')`. Then run:
```console
$ spack install m4  # extends libsigsegv
$ spack uninstall m4
```

### PR DESCRIPTION
**Pull request by adamjstewart**
Fixes #11980 

This is the simplest possible fix to the issue I raised in #11980. Currently, if you have an external package like `jdk` which happens to be extendable, and you install an extension for that package, it becomes impossible to install that package or its extensions if you don't have write access to that directory. For example, on macOS, JDK is installed to the system `/Library` directory. It looks like Spack is trying to create a file to keep track of which extensions are activated, but crashes because that is not possible.

Tested on Python 2 and 3.

### CODE REVIEW

**Comment by tgamblin** [comment]
@becker33: can you take a look at this one for triage?  It's view-related.

File: lib/spack/spack/filesystem_view.py [hunk 1]
```diff
@@ -201,9 +201,12 @@ def __init__(self, root, layout, **kwargs):
...
                 # projection but it makes sense for consistency
-                mkdirp(os.path.dirname(projections_path))
-                with open(projections_path, 'w') as f:
-                    f.write(s_yaml.dump({'projections': self.projections}))
+                try:
+                    mkdirp(os.path.dirname(projections_path))
+                    with open(projections_path, 'w') as f:
+                        f.write(s_yaml.dump({'projections': self.projections}))
+                except OSError:
+                    pass
```
Comment (becker33) [inline lib/spack/spack/filesystem_view.py]: This should raise when `self.projections` is non-empty.

Maybe we should do without the file entirely when `self.projections` is empty, but that could be a separate PR optimization as well.

Since extensions are activated using views with the default projections, raising `if self.projections` should not effect the bugfix in this PR.

```
except OSError as e:
    if self.projections:
        raise e
```

Comment (adamjstewart) [reply lib/spack/spack/filesystem_view.py]: Done. I confirmed that I am still able to uninstall extensions with this change. Other than that, I really don't know much about this file.

**Comment by adamjstewart** [comment]
Ping @becker33
